Fix jitMemcpy condition in copyCompactAndLinkCode

justinmichaud opened the pull request:
#### b1f668b135a35ba2e21118b3a0228338fb8252d3
<pre>
Fix jitMemcpy condition in copyCompactAndLinkCode
<a href="https://bugs.webkit.org/show_bug.cgi?id=298714">https://bugs.webkit.org/show_bug.cgi?id=298714</a>

Reviewed by Keith Miller.

In the case when r^x is enforced, but fast jit permissions are not
supported, we should use jitMemcpy.

Canonical link: <a href="https://commits.webkit.org/300773@main">https://commits.webkit.org/300773@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4dd92f41866c7f7a1f60c138e169e69bec2faf3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122063 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41765 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32435 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128626 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74157 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42480 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50359 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92778 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61661 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125015 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33879 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109309 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73434 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32892 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27475 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72120 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/114211 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103386 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131387 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120589 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49002 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37270 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101340 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49376 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101211 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26047 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46577 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24693 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45737 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48859 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54593 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150747 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48329 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38570 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51679 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50009 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/assembler/LinkBuffer.cpp:
```diff
@@ -413,20 +413,20 @@ void LinkBuffer::copyCompactAndLinkCode(MacroAssembler& macroAssembler, JITCompi
-        if (shouldCopyDirectlyToJITRegion)
-            MacroAssembler::link<memcpyRepatch>(linkRecord, outData + linkRecord.from(), location, target);
-        else
+        if (shouldCopyDirectlyToJITRegion && !g_jscConfig.useFastJITPermissions)
```
m_shouldPerformBranchCompaction && useFastJITPermissions: memcpy to jit region
!m_shouldPerformBranchCompaction && useFastJITPermissions: memcpy to jit region
m_shouldPerformBranchCompaction && !useFastJITPermissions: memcpy to buffer
!m_shouldPerformBranchCompaction && !useFastJITPermissions: jitmemcpy to jit region

justinmichaud replied:
Also: JITMemcpy should not be used when we aren't copying to a jit region IMO, for clarity.

kmiller68 replied:
Ah, I see. Yeah, that seems reasonable/correct, AFAICT.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a7ca9538ef5bf6cbbc1107789c94046cac982a9e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50596

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4dd92f41866c7f7a1f60c138e169e69bec2faf3b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50596

kmiller68 reviewed (approved):
r=me

justinmichaud committed:
Fix jitMemcpy condition in copyCompactAndLinkCode
https://bugs.webkit.org/show_bug.cgi?id=298714

Reviewed by Keith Miller.

In the case when r^x is enforced, but fast jit permissions are not
supported, we should use jitMemcpy.

Canonical link: https://commits.webkit.org/300773@main
